owlkettle/playground.nim

```diff
@@ -32,71 +32,72 @@ macro getField*(someType: untyped, fieldName: static string): untyped =
   nnkDotExpr.newTree(someType, ident(fieldName))
 
 # Default `toFormField` implementations
-proc toFormField[T: SomeNumber](state: auto, fieldName: static string, typ: typedesc[T]): Widget =
+proc toFormField(state: auto, field: ptr SomeNumber, fieldName: static string): Widget =
```

Comment:
- The `state` parameter is not used anymore.
- `fieldName` does not have to be static

Reply:
state is necessary to spawn dialogues, so for some of the overloads (only DateTime).
Thus the parameter is mandatory.

FieldName I fixed though. Also added a delete button to the toFormField for seq's

Reply:
Ah, I see. It should be possible to make it of type `Viewable` instead of `auto` though? (so, the compiler does not generate generic overloads for each widget `toFormField` is used in).

Reply:
Funnily enough, no, for the same reason as before.
Just tried it out.
Using Viewable *will* blow up the list_view example

Reply:
I think the issue here is that `open` has not yet been defined, as it is defined in `owlkettle.nim` which is not imported here. There are two options to fix this: import ../owlkettle (check if its recursive though) or move `open` to another file (e.g. `mainloop`).

Reply:
playground.nim can import owlkettle.nim without issue, however that does not appear to be a direct solution.
Even split out into its own file like this:
```nim
import ./widgets
import ./widgetdef
import./bindings/gtk

proc open*(app: Viewable, widget: Widget): tuple[res: DialogResponse, state: WidgetState] =
  let
    state = widget.build()
    dialogState = state.unwrapRenderable()
    window = app.unwrapInternalWidget()
    dialog = state.unwrapInternalWidget()
  gtk_window_set_transient_for(dialog, window)
  gtk_window_set_modal(dialog, cbool(bool(true)))
  gtk_window_present(dialog)
  
  proc destroy(dialog: GtkWidget, closed: ptr bool) {.cdecl.} =
    closed[] = true
  
  var closed = false
  discard g_signal_connect(dialog, "destroy", destroy, closed.addr)
  
  if dialogState of DialogState or dialogState of BuiltinDialogState:
    proc response(dialog: GtkWidget, responseId: cint, res: ptr cint) {.cdecl.} =
      res[] = responseId
    
    var res = low(cint)
    discard g_signal_connect(dialog, "response", response, res.addr)
    while res == low(cint):
      discard g_main_context_iteration(nil.GMainContext, cbool(ord(true)))
    
    state.read()
    if not closed:
      gtk_window_destroy(dialog)
    result = (toDialogResponse(res), state)
  else:
    while not closed:
      discard g_main_context_iteration(nil.GMainContext, cbool(ord(true)))
    
    state.read()
    result = (DialogResponse(), state)
```
Leads to errors
```nim
/home/philipp/dev/owlkettle/owlkettle/playground.nim(140, 10) template/generic instantiation of `gui` from here
/home/philipp/dev/owlkettle/owlkettle/playground.nim(148, 41) Error: type mismatch
Expression: open(state, block:
  let :tmp_2080375130 = DateDialog()
  :tmp_2080375130)
  [1] state: Widget
  [2] block:
  let :tmp_2080375130 = DateDialog()
  :tmp_2080375130: DateDialog

Expected one of (first mismatch at [position]):
[1] proc open(app: Viewable; widget: Widget): tuple[res: DialogResponse,
    state: WidgetState]
[1] proc open(f: var File; filehandle: FileHandle; mode: FileMode = fmRead): bool
[1] proc open(f: var File; filename: string; mode: FileMode = fmRead;
          bufSize: int = -1): bool
[1] proc open(filename: string; mode: FileMode = fmRead; bufSize: int = -1): File
[1] proc open[TMsg](c: var Channel[TMsg]; maxItems: int = 0)
```

Reply:
I just tried this and it works fine for me. Not sure why you get type errors.

Reply:
Ugh, classic case of me having further changes that lead to this because I was going through multiple different approaches on solving this.
Now this leaves on what to do regarding open().
It's effectively just a proc to open a Dialog in general right?
(So if I were to name it: `openDialog`).

Now the question mostly becomes where to put it.
I don't really have a full understanding of the module split (Note: module comments that declare the "Scope" of a given module would be nice).
Honestly as a helper proc for dialogue I'd have put it in widgets.nim
However that appears to not want to work come hell or high heaven, it keeps finding type mismatches :
```nim
Error: type mismatch
Expression: g_signal_connect(dialog, "destroy", destroy, addr(closed))
  [1] dialog: GtkWidget
  [2] "destroy": string
  [3] destroy: proc (dialog: GtkWidget, closed: ptr bool){.cdecl, noSideEffect, gcsafe.} | proc (ctx: CairoContext){.inline, noSideEffect, gcsafe.} | proc (surface: CairoSurface){.inline, noSideEffect, gcsafe.} | proc (pattern: CairoPattern){.inline, noSideEffect, gcsafe.}
  [4] addr(closed): ptr bool

Expected one of (first mismatch at [position]):
[1] proc g_signal_connect(app: GApplication; signal: cstring; closure, data: pointer): culong
[1] proc g_signal_connect(app: GtkListItemFactory; signal: cstring;
                      closure, data: pointer): culong
[1] proc g_signal_connect(app: GtkSelectionModel; signal: cstring;
                      closure, data: pointer): culong
[1] proc g_signal_connect(widget: GtkEventController; signal: cstring;
                      closure, data: pointer): culong
[3] proc g_signal_connect(widget: GtkWidget; signal: cstring; closure, data: pointer): culong
```
And it makes no sense to me that 3 can not be turned into a ptr when the proc is defined in that module.
Putting it in mainloop seems kinda random, as seems making its own module. It doesn't feel like the kind of place I'd intuitively look for this kind of thing/fits into the way things have been structured so far.

Reply:
I fixed it. For some reason having the destroy-proc named "destroy" was something nim did *not* like.
Simple renaming it to `destroyDialog` fixed it.

I moved `open` to widgets.nim since I believe it is most appropriate there as a proc required by a widget to function.
Refactored the entire thing to use Viewable and pushed it just now.